File: 0121-best-time-to-buy-and-sell-stock-买卖股票的最佳时机/0121.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if (prices.size() < 2)return 0;
        int buy = prices[0];
        int price = 0;
        for(int i = 1; i != prices.size(); ++i) {
            buy = prices[i] < buy ? prices[i] : buy;
            price = prices[i] - buy > price ? prices[i] - buy : price;
        }
        return price;
    }
};


/*
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if (prices.empty()) return 0;
        int buy = prices[0];
        int res = 0;
        for (int i = 0; i < prices.size(); ++i) {
            buy = min(buy, prices[i]);
            res = max(res, prices[i] - buy);
        }
        return res;
    }
};
*/

File: 0121-best-time-to-buy-and-sell-stock-买卖股票的最佳时机/0121.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        buy = 0
        sell = 0
        profit = 0
        while(sell < len(prices) - 1):
            sell += 1
            curProfit = prices[sell] - prices[buy]
            if curProfit < 0:
                buy = sell
                continue
            if curProfit > profit:
                profit = curProfit
        return profit